docs/proposals/scheduling/replicas-assign/divide-replicas-by-static-weight-evenly.md

```diff
@@ -0,0 +1,220 @@
+---
+title: divide replicas by static weight evenly
+authors:
+- "@chaosi-zju"
+reviewers:
+- "@RainbowMango"
+- "@XiShanYongYe-Chang"
+- "@zhzhuang-zju"
+approvers:
+- "@RainbowMango"
+
+creation-date: 2023-11-11
+---
+
+# Divide replicas by static weight evenly
+
+## Summary
+
+Karmada supports specifying the number of replicas to be allocated to each member cluster in a static weight ratio,
+e.g., the user specifies that the total replicas of a Deployment is 5,
+and the requirement is to distribute them into two clusters, member1 and member2, with a static weight of 1:1.
+
+According to the current implementation, the result must be 3 replicas for member1 and 2 replicas for member2.
+(5 is an odd number, so it can't be divided equally, and member1 allocates one more).
+
+This is because for clusters with equal weights, the remainder prefers the cluster with the smaller dictionary order of cluster names 
+in current implementation (the remainder is the number of replicas left over if the total replicas is not exactly divided by the static weights).
+
+If there are many such Deployments, the total replicas of member1 cluster will be significantly higher than that of member2 cluster, 
+i.e., a potential problem of uneven replicas distribution.
+
+Therefore, in order to solve the uneven replicas allocation problem without affecting the scheduling inertia, 
+I propose to optimize the sorting of clusters when allocating the remainder as:
+
+* Sort the clusters first by weights
+* If the weights are equal, then the clusters are sorted according to the current number of replicas
+   (more current number of replicas implies that the remainders of the last scheduling were randomized to such clusters,
+   and in order to keep the inertia in this scheduling, such clusters should also be prioritized).
+* If the weights are equal and the current number of replicas is equal too, then the clusters will be randomized.
```

Comment:
First of all, I think this is a good proposal with valuable optimization point. But the random algorithm is uncertain, and we cannot rely on it to ensure an even distribution of replicas. I think it would be better to base on the available replicas per cluster.
For `E.g.1 Expanded replicas scenario`:
Let us make an extreme assumption. If member2 is randomly selected every time, the replicas distributed on member2 will be significantly different from member3 and member4.

Reply:
>  I think it would be better to base on the available replicas per cluster.

base on available resources is the `dynamicWeight` scenario, this proposal focus on static weight.

> Let us make an extreme assumption. If member2 is randomly selected every time, the replicas distributed on member2 will be significantly different from member3 and member4.

This may not happen, if `member2`/`member3`/`member4` weight equally, each cluster will be randomly selected.

Reply:
@whitewindmills What you said makes sense. If the number of deployments is little, there actually exist this extremely assumption, but at this point the `remainder` is little too, may lead to little bad influence. 

However, if the number of deployments is huge, I think a simple random function can also do great job, it can achieve randomization.

So this is a compromise measure, and a small amount of modification can do effective work. Basing on the available replicas per cluster would indeed bring more benefits, but not so much.